src/xform_to_datalog/arcs_manifest_tree/particle_spec.cc

```diff
@@ -0,0 +1,80 @@
+#include "src/xform_to_datalog/arcs_manifest_tree/particle_spec.h"
+
+namespace raksha::xform_to_datalog::arcs_manifest_tree {
+
+namespace ir = raksha::ir;
+
+ParticleSpec ParticleSpec::CreateFromProto(
+    const arcs::ParticleSpecProto &particle_spec_proto) {
+  std::string name = particle_spec_proto.name();
+  CHECK(!name.empty()) << "Expected particle spec to have a name.";
+
+  std::vector<HandleConnectionSpec> handle_connection_specs;
+  for (const arcs::HandleConnectionSpecProto &hcs_proto :
+      particle_spec_proto.connections()) {
+    handle_connection_specs.push_back(
+        HandleConnectionSpec::CreateFromProto(hcs_proto));
+  }
+
+  std::vector<ir::TagClaim> claims;
+  for (const arcs::ClaimProto &claim : particle_spec_proto.claims()) {
+    CHECK(!claim.has_derives_from())
+      << "DerivesFrom claims not yet implemented";
+    CHECK(claim.has_assume()) << "Expected claim to carry an Assume field.";
+    claims.push_back(ir::TagClaim::CreateFromProto(claim.assume()));
+  }
+
+  std::vector<ir::TagCheck> checks;
+  for (const arcs::CheckProto &check : particle_spec_proto.checks()) {
+    checks.push_back(ir::TagCheck::CreateFromProto(check));
+  }
+
+  std::vector<ir::Edge> edges =
+      GenerateEdges(name, std::move(handle_connection_specs));
+
+  return ParticleSpec(
+      std::move(name), std::move(checks), std::move(claims), std::move(edges));
+}
+
+std::vector<ir::Edge> ParticleSpec::GenerateEdges(
+    std::string particle_spec_name,
```

Comment:
Just an observation: I still cannot get used to value parameters.

Reply:
All of my old C++ instincts scream "You're making an unnecessary copy! Why would you do that?". I'm pretty unnerved as well.